File: backend/src/test/java/com/just/donate/models/DonationPartTest.java
package com.just.donate.models;


class DonationPartTest {

    /*
    @Test
    void testBasisDonationPart() {
        DonationPart donationPart = new DonationPart(BigDecimal.valueOf(100), LocalDateTime.now());
        assertEquals(donationPart.getAmount(), BigDecimal.valueOf(100));

        DonationPart donationPart2 = new DonationPart(BigDecimal.valueOf(0), LocalDateTime.now());
        assertEquals(donationPart2.getAmount(), BigDecimal.valueOf(0));

        DonationPart donationPart3 = new DonationPart(BigDecimal.valueOf(-100), LocalDateTime.now());
        assertEquals(donationPart3.getAmount(), BigDecimal.valueOf(-100));
    }

    @Test
    void testSplitNothingOf() {
        DonationPart donationPart = new DonationPart(BigDecimal.valueOf(100), LocalDateTime.now());
        Splittable.Split<DonationPart, BigDecimal> split = donationPart.splitOf(BigDecimal.valueOf(0));

        assertTrue(split.getSplit().isEmpty());
        assertTrue(split.getOpen().isEmpty());
        assertTrue(split.getRemain().isPresent());
        assertEquals(split.getRemain().get().getAmount(), BigDecimal.valueOf(100));
        assertTrue(split.fullRemain());
    }

    @Test
    void splitALittleOf() {
        DonationPart donationPart = new DonationPart(BigDecimal.valueOf(100), LocalDateTime.now());
        Splittable.Split<DonationPart, BigDecimal> split = donationPart.splitOf(BigDecimal.valueOf(50));

        assertTrue(split.getSplit().isPresent());
        assertEquals(split.getSplit().get().getAmount(), BigDecimal.valueOf(50));
        assertTrue(split.getOpen().isEmpty());
        assertTrue(split.getRemain().isPresent());
        assertEquals(split.getRemain().get().getAmount(), BigDecimal.valueOf(50));
        assertTrue(split.someSplit());
    }

    @Test
    void testSplitOfExact() {
        DonationPart donationPart = new DonationPart(BigDecimal.valueOf(100), LocalDateTime.now());
        Splittable.Split<DonationPart, BigDecimal> split = donationPart.splitOf(BigDecimal.valueOf(100));

        assertTrue(split.getSplit().isPresent());
        assertEquals(split.getSplit().get().getAmount(), BigDecimal.valueOf(100));
        assertTrue(split.getOpen().isEmpty());
        assertTrue(split.getRemain().isEmpty());
        assertTrue(split.fullSplit());
    }

    @Test
    void testSplitOfMore() {
        DonationPart donationPart = new DonationPart(BigDecimal.valueOf(100), LocalDateTime.now());
        Splittable.Split<DonationPart, BigDecimal> split = donationPart.splitOf(BigDecimal.valueOf(150));

        assertTrue(split.getSplit().isPresent());
        assertEquals(split.getSplit().get().getAmount(), BigDecimal.valueOf(100));
        assertTrue(split.getOpen().isPresent());
        assertEquals(split.getOpen().get(), BigDecimal.valueOf(50));
        assertTrue(split.fullOpenSplit());
    }

    @Test
    void testSplitEmpty() {
        DonationPart donationPart = new DonationPart(BigDecimal.valueOf(0), LocalDateTime.now());
        assertThrows(IllegalStateException.class, () -> donationPart.splitOf(BigDecimal.valueOf(100)));
    }

        */
}

File: backend/src/test/scala/com/just/donate/models/DonationPartTest.scala
package com.just.donate.models

import org.scalatest.Assertions.*
import org.scalatest.flatspec.AnyFlatSpec
import org.scalatest.matchers.should.Matchers

import java.time.LocalDateTime

class DonationPartTest extends AnyFlatSpec with Matchers:

  "A DonationPart" should "correctly initialize with given amounts" in {
    val donationPart = Donation("Donor", BigDecimal(100))
    donationPart.amount shouldEqual BigDecimal(100)

    val donationPart2 = Donation("Donor", BigDecimal(0))
    donationPart2.amount shouldEqual BigDecimal.valueOf(0)

    val donationPart3 = Donation("Donor", BigDecimal(-100))
    donationPart3.amount shouldEqual BigDecimal.valueOf(-100)
  }

  it should "handle splitting with zero amount correctly" in {
    val donationPart = Donation("Donor", BigDecimal(100))
    val split = donationPart.splitOf(BigDecimal.valueOf(0))

    split.split shouldBe empty
    split.open shouldBe empty
    split.remain shouldBe defined
    split.remain.get.amount shouldEqual BigDecimal.valueOf(100)
  }

  it should "split a partial amount correctly" in {
    val donationPart = Donation("Donor", BigDecimal(100))
    val split = donationPart.splitOf(BigDecimal.valueOf(50))

    split.split shouldBe defined
    split.split.get.amount shouldEqual BigDecimal.valueOf(50)
    split.open shouldBe empty
    split.remain shouldBe defined
    split.remain.get.amount shouldEqual BigDecimal.valueOf(50)
  }

  it should "split the exact amount correctly" in {
    val donationPart = Donation("Donor", BigDecimal(100))
    val split = donationPart.splitOf(BigDecimal.valueOf(100))

    split.split shouldBe defined
    split.split.get.amount shouldEqual BigDecimal.valueOf(100)
    split.open shouldBe empty
    split.remain shouldBe empty
  }

  it should "handle splitting more than available amount correctly" in {
    val donationPart = Donation("Donor", BigDecimal(100))
    val split = donationPart.splitOf(BigDecimal.valueOf(150))

    split.split shouldBe defined
    split.split.get.amount shouldEqual BigDecimal.valueOf(100)
    split.open shouldBe defined
    split.open.get shouldEqual BigDecimal.valueOf(50)
  }

  it should "throw an IllegalStateException when splitting from an empty donation part" in {
    val donationPart = Donation("Donor", BigDecimal(0))
    an[IllegalStateException] should be thrownBy {
      donationPart.splitOf(BigDecimal.valueOf(100))
    }
  }
